modules/expandable-panel/ExpandablePanel.tsx

```diff
@@ -0,0 +1,229 @@
+  const bodyRef = useRef<HTMLDivElement>(null);
+
+  // calculate header height (once or every time window size changes)
+  useEffect(() => {
+    setHeaderHeight(headerRef.current?.getBoundingClientRect().height);
+  }, [headerRef, windowSize]);
+
+  // calculate body height (once or every time window size changes)
+  useEffect(() => {
+    if (bodyHeight === 0 || bodyHeight === undefined) {
+      setBodyHeight(bodyRef.current?.getBoundingClientRect().height);
+    }
+  }, [bodyHeight, bodyRef, isExpanded, windowSize]);
+
+  const handleCollapse = useCallback(() => {
+    setCurrentBodyHeight(0);
+    setIsCollapsing(true);
+
+    // allow the animation to finish before removing the body from the DOM
+    setTimeout(() => {
+      setIsCollapsing(false);
+    }, SLIDE_ANIMATION_DURATION_MS);
+  }, []);
+
+  // set body height manually to be able to have the sliding animation
+  useEffect(() => {
+    if (isExpanded && bodyHeight && bodyHeight > 0 && currentBodyHeight === 0) {
+      setTimeout(() => {
+        setCurrentBodyHeight(bodyHeight);
+      }, 300);
+    }
+  }, [bodyHeight, currentBodyHeight, isExpanded]);
+
+  // state is being controlled by the parent component
+  useEffect(() => {
+    if (expanded !== undefined) {
+      if (!expanded) {
+        handleCollapse();
+      }
+
+      setIsExpanded(expanded);
+    }
+  }, [expanded, handleCollapse]);
+
+  const handleClick = () => {
+    if (disabled) {
+      return;
+    }
+
+    // state is being controlled internally
+    if (expanded === undefined) {
+      if (isExpanded) {
+        handleCollapse();
+      }
+
+      setIsExpanded(!isExpanded);
+    }
+
+    // eslint-disable-next-line mdx/no-unused-expressions
+    onClick?.();
+  };
+
+  const stateClasses = useMemo(() => {
+    const classes = [];
+
+    if (disabled) {
+      classes.push("disabled");
+    }
+
+    if (isExpanded) {
+      classes.push("expanded");
+    }
+
+    return classes;
+  }, [disabled, isExpanded]);
+
+  const renderHeader = useCallback(
+    () => (
+      <div ref={headerRef} className={cx(styles.header, ...stateClasses)}>
+        {typeof header === "string" ? (
+          <TextHighlight className={cx(styles.headerTitle)}>
+            {header}
+          </TextHighlight>
+        ) : (
+          header?.(isExpanded)
+        )}
+        <Icon
+          className={cx(styles.headerIcon, ...stateClasses)}
+          name="arrow"
+          size={16}
+        />
+      </div>
+    ),
+    [header, cx, styles, stateClasses, isExpanded]
+  );
+
+  const isBodyHeightReady = bodyHeight && bodyHeight > 0;
+  const canAnimate = isBodyHeightReady && (isExpanded || isCollapsing);
+  const finalBodyHeight = canAnimate ? currentBodyHeight : "100%";
+  const bodyWrapperStyles = {
+    height: finalBodyHeight,
+    maxHeight: finalBodyHeight === 0 && !isCollapsing ? 0 : undefined
+  };
+
+  return (
+    <div
+      className={cx(className, styles.panel, ...stateClasses)}
```

Comment:
instead of having this `...stateClasses` throught the different pieces of components. we could have a selector that does thit for us. Something like: 
```javascript
const stylesheet: ThemeStyleSheetFactory = (theme: Theme) => ({
  panel: {
    overflow: "hidden",
    "@selectors": {
         "&.disabled .header" {
                ...apply styles for header here
          }
   }
  }
});
```

Reply:
I think the pattern we have adhered to is to apply the modifier class to the necessary elements instead of your suggestion. For consistency's sake, I will leave as is and we can discuss as a team whether we want to change this pattern.